pipeline-validation/custom-pipeline-validate-orca/src/main/java/com/opsmx/plugin/pipeline/validate/runtime/PipelineValidateAgentPreprocessor.java

```diff
@@ -0,0 +1,114 @@
+		logger.debug("ExecutionPreprocessor Type :{}",type);
+		if(type.equals(Type.PIPELINE)){
+			return true;
+		}
+		return false;
+	}
+	@Override
+	public Map<String, Object> process(@Nonnull Map<String, Object> pipeline){
+		logger.debug("Start of the Pipeline Validation");
+		logger.debug("input Pipeline :{}", pipeline);
+		if (!isPipelineValidateEnabled) {
+			logger.debug("Pipeline validation not enabled, returning");
+			logger.debug("End of the Pipeline Validation");
+			return pipeline;
+		}
+		try {
+           List stages = (ArrayList)pipeline.get("stages");
+		   List<String> refIds = new ArrayList<String>();
+			List<String> allRequisiteStageRefIds = new ArrayList<String>();
+		   for(Object stageObject : stages){
+			   Map<String, Object> stage = (HashMap) stageObject;
+			   refIds.add(stage.get("refId").toString());
+			   List<String> requisiteStageRefIds = (ArrayList)stage.get("requisiteStageRefIds");
+			   allRequisiteStageRefIds.addAll(requisiteStageRefIds);
+		   }
+           validateRefIds(refIds);
+		   validateRequisiteStageRefIds(refIds,allRequisiteStageRefIds);
+		   validateRefIdAndRequisiteStageRefIds(stages);
+		} catch (Exception e) {
+			e.printStackTrace();
+			//logger.error("Exception occured : {}", e);
+			logger.debug("Some thing wrong While validate pipeline input: {}", pipeline);
+			logger.debug("End of the Pipeline Validation");
+			throw new ValidationException(e.getMessage(), null);
+		}
+		logger.debug("End of the Pipeline Validation");
+		return pipeline;
+	}
+
+	private void validateRefIds(List<String> refIds) {
+		TreeMap<String, Integer> tmap = new TreeMap<String, Integer>();
+		for (String refId : refIds) {
+			Integer count = tmap.get(refId);
+			tmap.put(refId, (count == null) ? 1 : count + 1);
+		}
+
+		for (Map.Entry m : tmap.entrySet()) {
+			logger.info("Frequency of " + m.getKey() + " is " + m.getValue());
+			if(Integer.valueOf(m.getValue().toString() )> 1){
+				logger.debug(" same stage ref Id  :{} used more than one time ",m.getKey());
+				throw new ValidationException("Pipeline has been used same stage ref Id  : "+m.getKey()+" more than one time, Update pipeline stage ref Id ", null);
+			}
+		}
+	}
+
+	private void validateRequisiteStageRefIds(List<String> refIds, List<String> allRequisiteStageRefIds) {
+		for (String refId : allRequisiteStageRefIds) {
+         if(!refIds.contains(refId)){
+			 logger.debug("Pipeline stage ref Id  :{} is not available ",refId);
+			 throw new ValidationException("Pipeline stage ref Id  :"+refId+" is not available ", null);
+		 }
+		}
+	}
+
+	private void validateRefIdAndRequisiteStageRefIds(List stages) {
+		for(Object stageObject : stages){
+			Map<String, Object> stage = (HashMap) stageObject;
+			String refId = stage.get("refId").toString();
+			List<String> requisiteStageRefIds = (ArrayList)stage.get("requisiteStageRefIds");
+			if(requisiteStageRefIds.contains(refId)){
```

Comment:
follow my comment for prev two methods... without description / comment i too got confused which merthod does what.... anyways here are the messages, see which one fits here and which one for prev method - 
The requisiteStageRefIds property must not reference non existent refId values. Invalid id(s): 

The requisiteStageRefIds property must not contain self-references. Self-referenced id(s):

Reply:
updated logic to verify across all stages and also error message